src/appleseed/foundation/math/sampling/qmcsamplingcontext.h

```diff
@@ -332,55 +399,90 @@ inline void QMCSamplingContext<RNG>::compute_offset()
 }
 
 template <typename RNG>
-template <typename T>
-inline T QMCSamplingContext<RNG>::next2(Tag<T>)
+template <typename T, int M>
+inline T QMCSamplingContext<RNG>::next2(Tag<T, M>)
 {
-    return next2(Tag<Vector<T, 1>>())[0];
+    return next2(Tag<Vector<T, 1>, M>())[0];
 }
 
 template <typename RNG>
 template <typename T, size_t N>
-inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>>)
+inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>, M_Halton>)
 {
     Vector<T, N> v;
 
     assert(m_sample_count == 0 || m_instance < m_sample_count);
     assert(N == m_dimension);
     assert(N <= PrimeTableSize);
 
-    if (m_mode == QMCMode)
+    if (m_instance < PrecomputedHaltonSequenceSize)
     {
-        if (m_instance < PrecomputedHaltonSequenceSize)
-        {
-            for (size_t i = 0; i < N; ++i)
-            {
-                v[i] = static_cast<T>(PrecomputedHaltonSequence[m_instance * 4 + i]);
-                v[i] = rotate(v[i], static_cast<T>(m_offset[i]));
-            }
-        }
-        else
+        for (size_t i = 0; i < N; ++i)
         {
-            v[0] = radical_inverse_base2<T>(m_instance);
-            v[0] = rotate(v[0], static_cast<T>(m_offset[0]));
-
-            for (size_t i = 1; i < N; ++i)
-            {
-                v[i] = fast_radical_inverse<T>(i, m_instance);
-                v[i] = rotate(v[i], static_cast<T>(m_offset[i]));
-            }
+            v[i] = static_cast<T>(PrecomputedHaltonSequence[m_instance * 4 + i]);
+            v[i] = rotate(v[i], static_cast<T>(m_offset[i]));
         }
     }
     else
     {
-        for (size_t i = 0; i < N; ++i)
-            v[i] = rand2<T>(m_rng);
+        v[0] = radical_inverse_base2<T>(m_instance);
+        v[0] = rotate(v[0], static_cast<T>(m_offset[0]));
+
+        for (size_t i = 1; i < N; ++i)
+        {
+            v[i] = fast_radical_inverse<T>(i, m_instance);
+            v[i] = rotate(v[i], static_cast<T>(m_offset[i]));
+        }
     }
 
     ++m_instance;
 
     return v;
 }
 
+template <typename RNG>
+template <typename T, size_t N>
+inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>, M_RNG>)
+{
+    Vector<T, N> v;
+
+    assert(m_sample_count == 0 || m_instance < m_sample_count);
+    assert(N == m_dimension);
+    assert(N <= PrimeTableSize);
+
+    for (size_t i = 0; i < N; ++i)
+        v[i] = rand2<T>(m_rng);
+
+    ++m_instance;
+
+    return v;
+}
+
+template <typename RNG>
+template <typename T, size_t N>
+inline Vector<T, N> QMCSamplingContext<RNG>::next2(Tag<Vector<T, N>, M_CMJ>)
```

Comment:
No method should be specified here

Reply:
I think it make more sense to fallback here instead of next2(Vec<T, N>, M).
Because if we fallback to Halton in Vec<T, N>, M, then we assume this will never be called if we set the mode to RNG?

Reply:
If RNG is enabled, we explicitly call RNG specialization of next2

Reply:
Correct, but it's just in the code. I guess it's a matter of taste. Unless it's actually better to fallback in Vec<T, N>, M ?

Reply:
Okay; let's think we have another sampling strategy "awesome" that works only with "unicorn" type:) In this scenario one will have to define another "fallback" for the "awesome" sampling strategy;

You are right, it works as it should now, and my suggestion is not about your code - it's good, but for future programmers that probably would implement new sampling strategies as it was done before, so they would define a new excessive function; It's true since the code itself is quite complicated;

I suggest saying "If we don't have a good specialization for next2 call, we use Halton", you saying "If we don't have a good specialization for CMJ next2 call, we use Halton". That's the difference.

But yes, it's a matter of taste anyway.

Reply:
I agree with you that your solution is simpler when it comes to implement a new method :)